File: programming/problog/goed/boardconfiguration.pl
:- use_module(library(lists)).

width(2).
height(2).
pressable_color(red).
pressable_color(green).
pressable_color(blue).
pressable_color(yellow).

%%%%%%%%%%%%%%%%%
% RANDOM EVENTS
%%%%%%%%%%%%%%%%%
1/4::uniform_color(red,X,Y);1/4::uniform_color(green,X,Y);1/4::uniform_color(blue,X,Y);1/4::uniform_color(yellow,X,Y).

position(X,Y) :-
	width(Width),
	height(Height),
	between(0,Width,X),
	between(0,Height,Y).
	
block(Color,X,Y) :- 
	position(X,Y),
	uniform_color(Color,X,Y).
	

	
%%%%%%%%%%%%%%%%%
% GAME PREDICATES
%%%%%%%%%%%%%%%%%
initial_board(Board) :-
	create_board(Board).

create_board(Board) :-
	create_board(0,Board,[]).
create_board(Y,Board,BoardAcc) :-
	height(Height),
	Y < Height,
	create_row(Y,Row),
	NewBoardAcc = [Row|BoardAcc],
	NewY is Y + 1,
	create_board(NewY,Board,NewBoardAcc).
create_board(Y,Board,BoardAcc) :-
	height(Y),
	create_row(Y,Row),
	Board = [Row|BoardAcc].
	
create_row(Y,Row) :-
	create_row(0,Y,Row,[]).
create_row(X,Y,Row,RowAcc) :-
	width(Width),
	X < Width,
	block(Color,X,Y),
	append(RowAcc,[block(Color,X,Y)],NewRowAcc),
	NewX is X + 1,
	create_row(NewX,Y,Row,NewRowAcc).
create_row(X,Y,Row,RowAcc) :-
	width(X),
	block(Color,X,Y),
	append(RowAcc,[block(Color,X,Y)],Row).
	
evidence(block(blue,0,2)). evidence(block(red,1,2)). evidence(block(blue,2,2)).
evidence(block(red,0,1)). evidence(block(yellow,1,1)). evidence(block(yellow,2,1)).
evidence(block(red,0,0)). %evidence(block(blue,1,0)). %evidence(block(green,2,0)).

query(initial_board(Board)).

File: programming/problog/python/boardconfiguration.py
from __future__ import print_function
import csv

from problog.program import PrologString
from problog.engine import DefaultEngine
from problog.logic import Constant,Var,Term

p = PrologString("""
:- use_module(library(apply)).
:- use_module(library(lists)).

pressable_color(red).
pressable_color(green).
pressable_color(blue).
pressable_color(yellow).

board(Board) :-
    pressable_color(Color00),
    pressable_color(Color10),
    pressable_color(Color20),
    pressable_color(Color01),
    pressable_color(Color11),
    pressable_color(Color21),
    pressable_color(Color02),
    pressable_color(Color12),
    pressable_color(Color22),
    Board = [[block(Color02,0,2),block(Color12,1,2),block(Color22,2,2)],
             [block(Color01,0,1),block(Color11,1,1),block(Color21,2,1)],
             [block(Color00,0,0),block(Color10,1,0),block(Color20,2,0)]
             ],
    check_valid_board(Board).

check_valid_board(Board) :-
    check_horizontal(Board),
    check_vertical(Board).
    
check_horizontal([]).
check_horizontal([Row|Tail]) :-
    check_list(Row),
    check_horizontal(Tail).
    
check_vertical(Board) :-
    transpose(Board,Transpose),
    check_horizontal(Transpose).
    
check_list([]).
check_list([block(Color,_,_)|Tail]) :-
    check_list(Tail,[Color]).
check_list([],Blocks) :-
    length(Blocks,N),
    N < 3.
check_list([block(Color1,_,_)|Tail1],[Color2|Tail2]) :-
    Color1 = Color2,
    length([Color2|Tail2],N),
    N < 3,
    NewBlocks = [Color1|[Color2|Tail2]],
    check_list(Tail1,NewBlocks).
check_list([block(Color1,_,_)|Tail1],[Color2|Tail2]) :-
    Color1 \= Color2,
    length([Color2|Tail2],N),
    N < 3,
    NewBlocks = [Color1],
    check_list(Tail1,NewBlocks).

transpose(Ls, Ts) :-
	lists_transpose(Ls, Ts).

lists_transpose([], []).
lists_transpose([L|Ls], Ts) :-
	maplist(same_length(L), Ls),
	foldl(transpose_, L, Ts, [L|Ls], _).

transpose_(_, Fs, Lists0, Lists) :-
	maplist(list_first_rest, Lists0, Fs, Lists).
	
list_first_rest([L|Ls], L, Ls).

same_length([],[]).
same_length([_|L1],[_|L2]) :- same_length(L1, L2).
""")

def main():

    engine = DefaultEngine(label_all=True)

    db = engine.prepare(p)
    query_board = Term('board', None)
    a = engine.query(db,query_board)
    with open("3x3.csv", "wb") as f:
        writer = csv.writer(f)
        writer.writerows(a)

if __name__ == '__main__':
    main()